### PR TITLE
Add support for anyOf

### PR DESCRIPTION
**Pull request by strahius**
### Reasons for making this change

This change adds support for [anyOf](https://spacetelescope.github.io/understanding-json-schema/reference/combining.html#anyof). The logic is built in the `ArrayField` component and handles all the various supported types (number, string, array etc).

Areas that need improvement:
- The layout is maybe not the prettiest.
- UISchema is not supported for anyOf. The challenge here, as far as I understand it is that UISchema needs specific types defined in the JSONSchema, whereas with anyOf, the type is determined on the fly.
- Array with anyOf, cannot also have fixed items at the moment. 
- anyOf item of type object is not supported. null is also not supported, but I saw there is a pull request (#329) for that.

### Checklist

* [ ] **I'm updating documentation**
  - [ ] I've checked the rendering of the Markdown text I've added
  - [ ] If I'm adding a new section, I've updated the Table of Content
* [x] **I'm adding or updating code**
  - [x] I've added and/or updated tests
  - [ ] I've updated docs if needed
* [x] **I'm adding a new feature**
  - [x] I've updated the playground with an example use of the feature

### CODE REVIEW

**Comment by cyxou** [comment]
Why it hasn't been merged yet?

**Comment by jManji** [comment]
@cyxou I am wondering this myself... I am happy to re-work on it, if there are any comments, and update the branch (there's quite a few conflicts as ArrayField has been updated since my PR).

**Comment by inostia** [comment]
+1, my project requires this and is on hold until this gets merged.

**Comment by Natim** [comment]
> +1, my project requires this and is on hold until this gets merged.

Don't hesitate to use this branch in your project in the meantime.

**Comment by kkarimi** [comment]
+1 for merging

**Comment by Natim** [comment]
I guess the most important missing feature is the support of `uiSchema`. The anyOf feature is important to support but this PR doesn't support it yet at the standard we want for that lib.

We need at least to improve the UX of using anyOf as well as supporting the uiSchema.

Merging is easy but supporting an incomplete implementation of the spec in the lib is a path we are not prepared to take.

However this work is a great start and we are highly interesting in this, feel free to keep iterating and improving this Pull-Request and we will merge it as soon as we feel comfortable with maintaining it on the long term.

**Comment by jManji** [comment]
Thanks for the detailed response @Natim! I wasn't aware what was wrong with it, but after your explanation, it makes sense. I remember updating the `uiSchema` was quite challenging, as it does not support dynamically added types... But now that I see there's some interest in this PR, I will keep working on it, first by updating it with master, then trying to add support for `uiSchema`

File: src/components/fields/ArrayField.js [hunk 1]
```diff
@@ -228,35 +284,69 @@ class ArrayField extends Component {
...
   };
 
   onSelectChange = (value) => {
-    this.asyncSetState({items: value});
+    const newState = Object.assign({}, this.state, {formData: {items: value}});
+    this.asyncSetState(newState);
   };
 
+  anyOfOptions(anyOfItems) {
+    return anyOfItems.map(item => ({value: item.type, label: item.type}));
```
Comment (exchgr) [inline src/components/fields/ArrayField.js]: Should it be `label: item.title`?

**Comment by exchgr** [comment]
Will this be able to support inheritance? Suppose I have an array of widgets. A widget can be of several types: Button, Slider, and Select. Button and Slider have the same attributes, but Select allows the user to define what's in its list. 

The issue I'm running into is that they all need to be `"object"`s, and having multiple items in an `anyOf` array with the same `type` causes it to just select the first item with that `type`.

Here's an example of what I'm trying to do:

```JavaScript
items: {
  anyOf: [
    {
      type: "object",
      title: "Button",

      properties: {
        name: {
          type: "string",
          title: "Name"
        }
      }
    },

    {
      type: "object",
      title: "Slider",

      properties: {
        name: {
          type: "string",
          title: "Name"
        }
      }
    },

    {
      type: "object",
      title: "Select",

      properties: {
        name: {
          type: "string",
          title: "Name",
        },

        fields: {
          "$ref": "#/definitions/select_fields"
        }
      }
    }
  ]
}
```

File: src/components/fields/ArrayField.js (hunk 1 above)
Comment (jManji) [inline src/components/fields/ArrayField.js]: Good spot, thanks! I think it should, yes. Maybe `item.title || item.type`, in case it's not defined...

Comment (exchgr) [inline src/components/fields/ArrayField.js]: Sounds good to me!

**Comment by jManji** [comment]
@exchgr I see what you're saying. This was deliberate as I did not have a use case for multiple items with same type. I _think_ that part of code responsible is [this line](https://github.com/mozilla-services/react-jsonschema-form/pull/417/files#diff-049f0cb4d66526a998749db1e232f07cR338), that tries to find the selected item in the anyOf schema.

It could be _easily_ changed so that `title` is preferred instead `type`. But `title` is not an obligatory property and required by `object`. And it would break for cases where an `anyOf` schema has multiple types with the same title. I _think_ there is a simple custom change you can make if you want. If the `SelectWidget` uses the title as the identifier instead of the type, then it should just work...

I guess the problem lies in the fact that there is no unique identifier between `anyOf` items. Ideally I think it should not reference the items by type, but also by the order in the schema. Or maybe, now that I think about it, it could be just the order... Thanks for the comment, I will check whether order can be used, next time I have a look at it!

**Comment by exchgr** [comment]
@jManji Thanks for the quick response. Going by the array index may solve the issue. I was also thinking that it'd be useful to have a way for the dropdown to influence an attribute that indicates the type on the backend.

Building on my example, suppose we have a few classes on the backend (in this case, Rails). Here's an example with polymorphic associations emulating inheritance across multiple tables (aka Multiple Table Inheritance):

```ruby
class DeviceType < ApplicationRecord
  has_many :controls
  # we're also able to work with each control's "controllable" association directly, through some methods not defined in this example
end

class Control < ApplicationRecord
  belongs_to :device_type
  belongs_to :controllable, polymorphic: true
end

class Button < ApplicationRecord
  has_one :control, as: :controllable
end

class Slider < ApplicationRecord
  has_one :control, as: :controllable
end

class Select < ApplicationRecord
  has_one :control, as: :controllable
end
```

Then in the JSON schema:
```JavaScript
items: {
  anyOf: [
    {
      type: "object",
      title: "Button",

      properties: {
        name: {
          type: "string",
          title: "Name"
        },

        controllable_type: {
          enum: ["Button"]
        }
      }
    },

    {
      type: "object",
      title: "Slider",

      properties: {
        name: {
          type: "string",
          title: "Name"
        },

        controllable_type: {
          enum: ["Slider"]
        }
      }
    },

    {
      type: "object",
      title: "Select",

      properties: {
        name: {
          type: "string",
          title: "Name",
        },

        controllable_type: {
          enum: ["Select"]
        }

        fields: {
          "$ref": "#/definitions/select_fields"
        }
      }
    }
  ]
}
```

Since I actually need this for something I'm working on, I'm going to take a stab at it today.

File: src/components/fields/ArrayField.js [hunk 2]
```diff
@@ -228,35 +284,69 @@ class ArrayField extends Component {
...
+  }
+
+  setWidgetType(index, value) {
+    const {items} = this.state.formData;
+    const {registry} = this.props;
+    const {definitions} = registry;
+    const anyOfItemsSchema = this.getAnyOfItemsSchema();
+    const newItems = items.slice();
+    const foundItem = anyOfItemsSchema.find((element) => element.type === value);
```
Comment (glasserc) [inline src/components/fields/ArrayField.js]: Does this work with things like `$ref`?

File: src/components/fields/ArrayField.js [hunk 3]
```diff
@@ -481,6 +580,16 @@ class ArrayField extends Component {
     };
     has.toolbar = Object.keys(has).some(key => has[key]);
 
+    const selectWidget = anyOfItemsSchema ? (
+      <div className="form-group" style={{width: 120}}>
+        <SelectWidget
+          schema={{type: "integer", default: selectWidgetValue}}
+          id="select-widget-id"
```
Comment (glasserc) [inline src/components/fields/ArrayField.js]: The same ID shouldn't be used more than once in a DOM.

File: src/components/fields/ArrayField.js [hunk 4]
```diff
@@ -50,6 +50,7 @@ function DefaultArrayItem(props) {
     <div key={props.index} className={props.className}>
 
       <div className={props.hasToolbar ? "col-xs-9" : "col-xs-12"}>
+        {props.selectWidget}
```
Comment (glasserc) [inline src/components/fields/ArrayField.js]: It seems kind of strange for `selectWidget` to be part of the props, and not part of the template.

**Comment by b1r3k** [comment]
@glasserc Thanks for review. Argument for extending support for anyOf beyond array makes sense + it's very useful to me at this moment.

For example I would like to use this schema:

```
{
    "definitions": {
        "TPNode": {
            "title": "TPNode",
            "description": "TPNode",
            "type": "object",
            "properties": {
                "name": {
                    "type": "string"
                },
                "value": {
                    "anyOf": [
                        { "type": "string"},
                        {
                            "type": "array",
                            "items": {
                                "$ref": "#/definitions/TPNode"
                            }
                        }
                    ]
                }
            }
        }
    },
    "$ref": "#/definitions/TPNode"
};
```

[Here is how it should render](http://jeremydorn.com/json-editor/?schema=N4IgJgpgZglgdjALjA9nAziAXKAKgBQDkVJtRlEAbCbEA40gGnAnQGMAnGAB2TVvokazRAE9uNLCBQAjAFYQ2iEM24cUEjslZkQcAIYBbSeXGSQ6RFzgBzEAF9mAN32UAriZD64ogPJRsAG1TCVpLaztHEPN9Dg59URUQJAhDTBwQABIOaFoAYgB6SFgEPgwCwVJ7ewBdaurmbNypQuL4JFRyypp7IAA&value=N4IgdghgtgpiBcIQBoQDcIBsCud4G1RJYElUMc8kBfAXWqAA&theme=bootstrap2&iconlib=fontawesome4&object_layout=normal&show_errors=interaction)

@jManji are you intending to work this PR according to @glasserc suggestions any time soon?

File: src/components/fields/ArrayField.js (hunk 2 above)
Comment (b1r3k) [inline src/components/fields/ArrayField.js]: I've tried it & it doesn't support $ref

**Comment by jManji** [comment]
Sorry for late reply guys. I've been quite busy lately, and didn't have time to properly look at this. @b1r3k yes, I would very much like to rework on this, when I have some time, based on @glasserc's suggestions. And there's a couple of things that came up from previous conversations here. I would say I should have something in the next couple of weeks. 

@glasserc, thanks a lot for the detailed CR, appreciated! Just trying to answer some of your questions without touching the code yet:
> How come the logic for this is part of the ArrayField class? Just to simplify, I guess?

I think that was because I wanted to make use of the `ArrayField` functionality that is there already, like adding multiple properties, all the UI stuff that comes with it, etc. `anyOf` looks and behaves a bit like an array, so I thought I could fit it in there (like for instance, how `additionalItems` feature does). Maybe a new component type makes more sense? That seemed much more work I guess.

> I also found the UI a little clunky -- rather than having a "select" to choose what kind of object it is, why not use something like Bootstrap's "radio" buttons functionality?

You think radio is better for this case? How would you render lots of options? Will it not occupy too much space and make the form look a bit inconsistent? Not sure, but that's and easy change, if that's what we want.

> Also, why did you move items from arrayField.state.items to arrayField.state.formData.items?

That's a good question... I must have had a reason, I will let you know on my next commit 😄. If I see no reason, I will just move it back to `state.items`

File: playground/samples/index.js [hunk 5]
```diff
@@ -30,4 +31,5 @@ export const samples = {
   Files: files,
   Single: single,
   "Custom Array": customArray,
+  "Any of": anyOf
```
Comment (exchgr) [inline playground/samples/index.js]: In any case (😉), leaving "of" as lowercase is correct, even when you're using title case. http://grammar.yourdictionary.com/capitalization/rules-for-capitalization-in-titles.html

**Comment by exchgr** [comment]
Perhaps if this supports `uiSchema`, we could let the developer decide whether to use radios or selects.

**Comment by glasserc** [comment]
I like @exchgr's suggestion to support uischema. As a default, I'd propose this heuristic: if the anyOf options are only two or three, I think radio buttons are more convenient (just click the one you want); if they are four or more, a select is probably more respectful of screen real estate.

> Maybe a new component type makes more sense?

Could we maybe put it on SchemaField? That was my first instinct (coming from a classical OO background), since any schema could potentially be `anyOf`. On the other hand, that might make SchemaField too complicated. Maybe @n1k0 has an opinion?

**Comment by n1k0** [comment]
I also like the idea of using uiSchema to configure the type selector a lot, though I fear the added complexity. Personally, I'd probably iterate here. Good support for anyOf is already hard to get right :)

Otherwise I concur with @glasserc, `SchemaField` feels like the natural place to implement such support.

Side note, @jManji could you please try to install [prettier](https://www.npmjs.com/package/prettier) v0.22.0 and run:

```
$ prettier --jsx-bracket-same-line --trailing-comma es5 '{playground,src,test}/**/*.js' --write
```

on your current branch, then try to merge/rebase it with latest master? I also fear we're gonna eventually diffing and conflicting a lot if we postpone this operation too much. Thanks!

**Comment by jManji** [comment]
The challenge I see with `uiSchema` is that it is keyed by the property's key. You cannot just say, "I want `anyOf` items to be `textarea`" for instance. You have to define which `anyOf` item you refer to, and since we potentially want (currently this branch does not support it) multiple `anyOf` items of the same type, the `uiSchema` needs to probably reference the order/index of the `anyOf` item (?). I will definitely leave the `uiSchema` out of the scope of the next iteration. If we want, we can start by figuring out how a `uiSchema` for `anyOf` would look like.

@glasserc, @n1k0 Regarding moving code to `SchemaField`, it's been some time since I looked at the code, but currently this also feels challenging/weird to me. It will definitely make `SchemaField` more complicated as you said, and I think, it will need to have some kind of support for dynamically added widgets, like `ArrayField` has. But I am sure you know the codebase much better than me, so I will look into how this can be achieved.

> Side note, @jManji could you please try to install prettier v0.22.0 and run...

Yes, will do!

>... I also fear we're gonna eventually diffing and conflicting a lot if we postpone this operation too much. 

Yes, you are right, I know... I will try to work on this asap, going through every comment mentioned here and in the code review.

**Comment by b1r3k** [comment]
@jManji any news regarding this PR? I'm willing to take it further if you do not have time. In fact I've forked your solution and fixed few mistakes in order to use it in our internal tooling but it has meant to be workaround and the further we go with development the more crippled we are due to divergence from origin/master

**Comment by jManji** [comment]
@b1r3k I am very sorry but unfortunately I don't plan on working on it again, any time soon. I needed that feature for my work, but now we have shifted priorities. Please feel free to work on it!

I haven't checked in detail, but maybe #581 solves the same problem in a different way?

**Comment by danrot** [comment]
I have checked #581, but it looks to me like this is only built to be used in dropdown and selects. From what I've seen it doesn't allow you to build more complex objects than simple strings. 

@b1r3k Have you continued working on it? Anything worth sharing? We would also desperately need a feature like this, although my use case would probably be a `oneOf`, although `anyOf` should also be good enough :-)

**Comment by intersplice** [comment]
Is there any way to install this through npm as a node_module?

**Pull request closed without merging**